src/js/trove/iolib.js

```diff
@@ -0,0 +1,63 @@
+({
+  requires: [ ],
+  provides: {
+    shorthands: { },
+    values: {
+      "prompt": ["arrow", ["String"], "String"],
+      "wrap-input-test": ["arrow", ["String"], "String"]
+    },
+    aliases: { },
+    datatypes: { }
+  },
+  nativeRequires: ["readline"],
+  theModule: function(RUNTIME, NAMESPACE, uri, readline) {
+    function Input(mockIO = () => {}) {
+        return RUNTIME.pauseStack(function(restarter) {
+          const rl = readline.createInterface({ input: RUNTIME.stdin });
+          
+          // input does not need to display anything
+          questionPromise = new Promise(resolve => rl.question('', input => resolve(input)));
+          
+          mockIO(rl)
+          
+          questionPromise
+            .then(result => restarter.resume(RUNTIME.makeString(result)))
+            .catch(error => {
+              // TODO: write a test for this
+              // TODO: we should probably NOT fail this hard
+              restarter.error(RUNTIME.makeString(error));
+            })
+            .finally(_ => rl.close())
+        })
+    };
+
+    function Prompt(msg) {
+      RUNTIME.stdout(msg);
+      return Input();
+    };
+
+    function wrapInputTest(testInput) {
+      return Input(mockIO = (rl) => rl.write(testInput + '\n'));
```

Comment:
This is my least hacky attempt at a reasonable, repeatable test. That being said, baking it into the actual library like this and exposing it as a public interface does not seem right.

Maybe a test helper or a test only module? 

The only solutions I have come up with thus far require interfacing with JS.

Reply:
@blerner my biggest struggle here is figuring out a reasonable test strategy. What I have here certainly works and is less hacky than my first solution but still not as clean or comprehensive as what jest/mocha or any other "production" mocking library provides.

An ideal solution (with heavy influence from Ruby world) might look something like:

```
// set up mock
before_check:
   stdin = mock-stdin();

   // inject/queue-up/send/etc.
   stdin.inject("hello")
end

// execute test
check:
    input() is "hello"
end
```

Do we want to investigate this? It feels like scope creep but it also feels like a pre-requisite for this work? Thoughts?

Beyond that, thanks for the multiple rounds of reviews so far! I apologize if the PR was pretty messy to start with - thanks for your patience here, very much appreciated.

Reply:
Most likely, this will require writing a test harness in  JS, rather than purely in Pyret.  We have some tests that have been written this way, but I think they aren't quite what you want to emulate here.  I'll need to talk to Joe a bit and bounce ideas together with him to see what we thing you ought to build here.

The idea of a `before-check` block is interesting, but definitely out of scope for now.  There are a lot of gnarly questions that would need to be designed (e.g. when does `stdin` get "reset" in your example to the "real" `stdin`? or what should the order of execution be when you have `where` blocks inside functions that are called by `before_check` or other `check` blocks?  A Pyret program weaves together testing and normal running into one source file, in ways that "real test code" typically doesn't...)

Reply:
I think instead of `mock-input-test` for now, you probably just want to invoke your tests as `exec("echo theInput | node theTest")`, more or less.  Something similar to https://github.com/brownplt/pyret-lang/blob/anchor/tests-new/simple-output.test.js or https://github.com/brownplt/pyret-lang/blob/anchor/tests-new/check-blocks.test.js#L13-L26, which are a test harness (not on the current #horizon branch, though!) that executes Pyret programs and checks their standard-out; you could pretty easily pipe in the necessary standard-in.